File: play/shared/formation.ts
/** Formation Drill / Guard / March — cheap same-company adjacency auras. */

import { hexDistOddR } from './hex'
import type { CardSnapshot } from './army'
import type { GameState, UnitToken } from './types'

/** UV threshold for Formation aura beneficiaries (inclusive). */
export const FORMATION_UV_CAP = 3

export const FORMATION_DRILL = 'Formation Drill'
export const FORMATION_GUARD = 'Formation Guard'
export const FORMATION_MARCH = 'Formation March'

function catalogCard(
  state: GameState,
  cardId: string | null | undefined,
): CardSnapshot | null {
  if (!cardId) return null
  return state.cardCatalog?.[cardId] ?? null
}

function unitHasFormationKeyword(unit: UnitToken, keyword: string): boolean {
  const needle = keyword.toLowerCase()
  if ((unit.abilities ?? []).some((a) => a.toLowerCase() === needle)) return true
  return (unit.keywords ?? []).some(
    (k) => k.toLowerCase() === needle || k.toLowerCase().startsWith(`${needle} `),
  )
}

/** Printed UV for a token (from catalog). */
export function unitUv(state: GameState, unit: UnitToken): number | null {
  const uv = catalogCard(state, unit.cardId)?.uv
  return typeof uv === 'number' && Number.isFinite(uv) ? uv : null
}

/** True if this model can receive Formation auras (UV ≤ cap). */
export function isFormationCheapUnit(state: GameState, unit: UnitToken): boolean {
  const uv = unitUv(state, unit)
  return uv != null && uv <= FORMATION_UV_CAP
}

/** Same company: share officerCardId (officer's points at self). */
export function sameCompany(a: UnitToken, b: UnitToken): boolean {
  return (
    a.seat === b.seat &&
    !!a.officerCardId &&
    !!b.officerCardId &&
    a.officerCardId === b.officerCardId
  )
}

/**
 * Adjacent same-company ally that provides the given Formation keyword.
 * The beneficiary does not need the keyword — the provider does.
 */
export function hasAdjacentFormationProvider(
  state: GameState,
  unit: UnitToken,
  keyword: string,
): boolean {
  return state.units.some(
    (m) =>
      m.id !== unit.id &&
      sameCompany(m, unit) &&
      hexDistOddR(m, unit) === 1 &&
      unitHasFormationKeyword(m, keyword),
  )
}

/** +1 Move while adjacent to Formation March (UV ≤ 3). */
export function formationMarchBonus(state: GameState, unit: UnitToken): number {
  if (!isFormationCheapUnit(state, unit)) return 0
  if (!hasAdjacentFormationProvider(state, unit, FORMATION_MARCH)) return 0
  return 1
}

/** +1 Hit (lower hit need by 1) while adjacent to Formation Drill (UV ≤ 3). */
export function formationDrillHitBonus(state: GameState, unit: UnitToken): number {
  if (!isFormationCheapUnit(state, unit)) return 0
  if (!hasAdjacentFormationProvider(state, unit, FORMATION_DRILL)) return 0
  return 1
}

/**
 * +1 Toughness while defending near Formation Guard (UV ≤ 3).
 * Implemented as Harden-style −1 damage (floor 1), matching battle sim.
 */
export function formationGuardMitigation(state: GameState, unit: UnitToken): number {
  if (!isFormationCheapUnit(state, unit)) return 0
  if (!hasAdjacentFormationProvider(state, unit, FORMATION_GUARD)) return 0
  return 1
}

File: unity/CommandWarfare/Assets/Scripts/Core/State/Formation.cs
using CommandWarfare.Core.Combat;
using CommandWarfare.Core.Hex;
using CommandWarfare.Core.Types;
using CommandWarfare.Data;

namespace CommandWarfare.Core.State
{
    /// <summary>Port of play/shared/formation.ts — adjacent same-company auras.</summary>
    public static class Formation
    {
        public const int UvCap = 3;
        public const string Drill = "Formation Drill";
        public const string Guard = "Formation Guard";
        public const string March = "Formation March";

        public static bool SameCompany(UnitToken a, UnitToken b) =>
            a.Seat == b.Seat &&
            !string.IsNullOrEmpty(a.OfficerCardId) &&
            a.OfficerCardId == b.OfficerCardId;

        public static int UnitUv(CardDatabase cards, UnitToken unit)
        {
            var card = cards?.FindById(unit.CardId);
            return card != null && card.uv > 0 ? card.uv : int.MaxValue;
        }

        public static bool IsCheapUnit(CardDatabase cards, UnitToken unit) =>
            UnitUv(cards, unit) <= UvCap;

        public static int MarchBonus(GameState state, UnitToken unit, CardDatabase cards)
        {
            if (!IsCheapUnit(cards, unit)) return 0;
            return HasAdjacentProvider(state, unit, March) ? 1 : 0;
        }

        public static int DrillHitBonus(GameState state, UnitToken unit, CardDatabase cards)
        {
            if (!IsCheapUnit(cards, unit)) return 0;
            return HasAdjacentProvider(state, unit, Drill) ? 1 : 0;
        }

        public static int GuardMitigation(GameState state, UnitToken unit, CardDatabase cards)
        {
            if (!IsCheapUnit(cards, unit)) return 0;
            return HasAdjacentProvider(state, unit, Guard) ? 1 : 0;
        }

        /// <summary>Pack: melee +1 Hit while adjacent to ≥2 friendly Pack units.</summary>
        public static bool PackMeleeHitBonus(GameState state, UnitToken attacker)
        {
            if (!CombatKeywords.HasUnitAbility(attacker, "Pack")) return false;
            var origin = new HexCoord(attacker.Col, attacker.Row);
            var n = 0;
            foreach (var u in state.Units)
            {
                if (u.Id == attacker.Id || u.Seat != attacker.Seat) continue;
                if (!CombatKeywords.HasUnitAbility(u, "Pack")) continue;
                if (HexMath.Distance(origin, new HexCoord(u.Col, u.Row)) == 1)
                    n++;
            }
            return n >= 2;
        }

        static bool HasAdjacentProvider(GameState state, UnitToken unit, string keyword)
        {
            var origin = new HexCoord(unit.Col, unit.Row);
            foreach (var m in state.Units)
            {
                if (m.Id == unit.Id || !SameCompany(m, unit)) continue;
                if (HexMath.Distance(origin, new HexCoord(m.Col, m.Row)) != 1) continue;
                if (HasFormationKeyword(m, keyword)) return true;
            }
            return false;
        }

        static bool HasFormationKeyword(UnitToken unit, string keyword)
        {
            if (HasAbilityOrKeyword(unit.Abilities, keyword)) return true;
            return HasAbilityOrKeyword(unit.Keywords, keyword);
        }

        static bool HasAbilityOrKeyword(System.Collections.Generic.List<string> list, string keyword)
        {
            if (list == null) return false;
            var needle = keyword.ToLowerInvariant();
            foreach (var entry in list)
            {
                if (string.IsNullOrEmpty(entry)) continue;
                var lower = entry.ToLowerInvariant();
                if (lower == needle || lower.StartsWith(needle + " ")) return true;
            }
            return false;
        }
    }
}
